File: web/src/features/rooms/types/room.ts
import { User } from "./user";

export type Room = {
  id: string;
  name: string;
  emoji: string;
  // user_id
  createdBy: string;
  members: Member[];
};

export type RoomResponse = {
  id: string;
  name: string;
  emoji: string;
  created_by: string;
  members: {
    id: string;
    room_id: string;
    user: {
      id: string;
      name: string;
      icon_url: string;
    };
    total_amount: number;
  }[];
};

export const jsonToRoom = (json: RoomResponse): Room => {
  return {
    id: json.id,
    name: json.name,
    emoji: json.emoji,
    createdBy: json.created_by,
    members: json.members.map((member) => ({
      id: member.id,
      roomId: member.room_id,
      user: {
        id: member.user.id,
        name: member.user.name,
        iconUrl: member.user.icon_url,
      },
      totalAmount: member.total_amount,
    })) as Member[],
  };
};

export type Member = {
  id: string;
  roomId: string;
  user: User;
  totalAmount: number;
};

File: backend/domain/src/room.rs
mod member;

pub use member::{Member, MemberId};

pub mod payment;
pub use payment::{Payment, PaymentId};

use crate::UserId;

use uuid::Uuid;

#[derive(serde::Serialize, Clone, sqlx::Type, PartialEq, Debug)]
#[sqlx(transparent)]
pub struct RoomId(pub String);

impl Default for RoomId {
    fn default() -> Self {
        Self(Uuid::now_v7().to_string())
    }
}

#[derive(serde::Serialize, sqlx::FromRow)]
pub struct Room {
    pub id: RoomId,
    // TODO: 名前の長さに制限をかける
    pub name: String,
    // TODO: Charでいいな
    // TODO: emoji文字列のバリデーション。
    pub emoji: String,
    // TODO: ownerとかにするか？
    pub created_by: UserId,
    // TODO: Vec<UserId>でいいな
    // TODO: ユニークである必要がある
    #[sqlx(skip)]
    pub members: Vec<Member>,
}
